README.md
# SpatialReference
> A helper library for translating between spatial reference systems

[![npm version][npm-img]][npm-url]
[![build status][travis-img]][travis-url]

[npm-img]: https://img.shields.io/npm/v/spatialreference.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/spatialreference
[travis-img]: https://travis-ci.org/GeoXForm/spatialreference.svg?branch=master
[travis-url]: https://travis-ci.org/GeoXForm/spatialreference

## Usage

Basic installation: `npm install spatialreference`

Then to use the module: `var SR = require('spatialreference')`

### With a DB
To use this module with a db that can store or retrieve WKTs, it needs to support (or be wrapped to support) two functions:
```javascript
function getWKT (wkid, function (err, wkt) {
// do some stuff to get a wkt
callback(null, wkt)
})
```
```javascript
function insertWKT (wkid, wkt, function (err) {
// store the wkt in the db
callback(null)
})
```
Then initialize the SpatialReference object with a db in the options.db key
```javascript
var SR = require('spatialreference')
var db = require('koop-pgcache') // or your db that supports getWKT and insertWKT
var sr = new SR({db: db})
```

### With a logger
By default SpatialReference will log to the console, but you can optionally pass in a logger in the options key. It needs to support a function like:
```javascript
function log('level','message') {
// write a log!
}
```
```javascript
var SR = require('spatialreference')
var logger = require('winston')
var sr = new SR({logger: logger})
```

## API

### wkidToWkt
This method takes a spatial reference system well-known id and gets the well-known text. If the WKID is an esri code, then it will use the esri-proj-codes library. If the WKID is available in a passed in database it will get it from there. Else it will make a request to http://www.epsg.io, insert the WKT into the database and return it to the client. 

Example using [koop-pgcache](https://github.com/koopjs/koop-pgcache)
``` javascript
var SpatialReference = require('spatialreference')
var db = require('koop-pgcache')

var sr = new SpatialReference({db: db})

sr.wkidToWkt(2927, function (err, wkt) {
	if (err) return console.error(err)
	console.log(wkt) 
	// 'PROJCS["NAD83(HARN) / Washington South (ftUS)",GEOGCS["NAD83(HARN)",DATUM["NAD83_High_Accuracy_Regional_Network",SPHEROID["GRS 1980",6378137,298.257222101,AUTHORITY["EPSG","7019"]],TOWGS84[0,0,0,0,0,0,0],AUTHORITY["EPSG","6152"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4152"]],UNIT["US survey foot",0.3048006096012192,AUTHORITY["EPSG","9003"]],PROJECTION["Lambert_Conformal_Conic_2SP"],PARAMETER["standard_parallel_1",47.33333333333334],PARAMETER["standard_parallel_2",45.83333333333334],PARAMETER["latitude_of_origin",45.33333333333334],PARAMETER["central_meridian",-120.5],PARAMETER["false_easting",1640416.667],PARAMETER["false_northing",0],AUTHORITY["EPSG","2927"],AXIS["X",EAST],AXIS["Y",NORTH]]'
})
```

### esriLookup
This is a pure wrapper around the [esri-proj-codes library](https://github.com/esri/esri-proj-codes). It gets the esri format of the well-known text for a projection when given the well-known id.

```javascript
var SpatialReference = require('spatialreference')

var sr = new SpatialReference()

sr.esriLookup(2927, function (err, wkt) {
	if (err) return console.error(err)
	console.log(wkt) 
	// PROJCS["NAD_1983_HARN_StatePlane_Washington_South_FIPS_4602_Feet",GEOGCS["GCS_North_American_1983_HARN",DATUM["D_North_American_1983_HARN",SPHEROID["GRS_1980",6378137.0,298.257222101]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]],PROJECTION["Lambert_Conformal_Conic"],PARAMETER["False_Easting",1640416.666666667],PARAMETER["False_Northing",0.0],PARAMETER["Central_Meridian",-120.5],PARAMETER["Standard_Parallel_1",45.83333333333334],PARAMETER["Standard_Parallel_2",47.33333333333334],PARAMETER["Latitude_Of_Origin",45.33333333333334],UNIT["Foot_US",0.3048006096012192]]
})
```

## Acknowledgements
This module leans heavily on the spatial reference information saved in PostGIS and the data available at epsg.io. It would not be possible without those two resources. So thank you to all involved.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.3] - 2018-10-05
### Changed
* Bump @esri/proj-codes to ^2.2.0

## [1.1.2] - 2018-09-30
### Fixed
* Fix code flow leading to insert when already found in DB.

## [1.1.1] - 2018-07-24
### Fixed
* Remove async/await and rewrite with promises to support older versions of Node

## [1.1.0] - 2018-07-23
### Changed
* Altered flow: first look in DB, then ESRI codes, then epsg.io
* Use promises, async/await, and as result, compile into a distribution that allow for use in older Node versions

### Added
* Also check ESRI deprecated codes

## [1.0.4] - 2017-03-14
### Fixed
* Use correct translation for non HARD NAD83

## [1.0.3] - 2016-03-25
### Added
* Override wkid: 2264

## [1.0.2] - 2016-01-04
### Fixed
* Add corrections for known problematic wkts

## [1.0.1] - 2015-09-18
### Changed
* Change from spatialreference.org to epsg.io

[1.1.3]: https://github.com/koopjs/spatialreference/compare/v1.1.2...v1.1.3
[1.1.2]: https://github.com/koopjs/spatialreference/compare/v1.1.1...v1.1.2
[1.1.1]: https://github.com/koopjs/spatialreference/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/koopjs/spatialreference/compare/v1.0.4...v1.1.0
[1.0.4]: https://github.com/koopjs/spatialreference/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/koopjs/spatialreference/compare/v1.0.3...v1.0.2
[1.0.2]: https://github.com/koopjs/spatialreference/compare/v1.0.2...v1.0.1
[1.0.1]: https://github.com/koopjs/spatialreference/compare/v1.0.0...v1.0.1


test/e2e.js

const test = require('tape')
const SpatialReference = require('../src/index')
const wkidHash = new Map()
wkidHash.set(2927, 'PROJCS["NAD83(HARN) / Washington South (ftUS)",GEOGCS["NAD83(HARN)",DATUM["NAD83_High_Accuracy_Regional_Network",SPHEROID["GRS 1980",6378137,298.257222101,AUTHORITY["EPSG","7019"]],TOWGS84[-0.9956,1.9013,0.5215,0.025915,0.009426,0.011599,-0.00062],AUTHORITY["EPSG","6152"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4152"]],UNIT["US survey foot",0.3048006096012192,AUTHORITY["EPSG","9003"]],PROJECTION["Lambert_Conformal_Conic_2SP"],PARAMETER["standard_parallel_1",47.33333333333334],PARAMETER["standard_parallel_2",45.83333333333334],PARAMETER["latitude_of_origin",45.33333333333334],PARAMETER["central_meridian",-120.5],PARAMETER["false_easting",1640416.667],PARAMETER["false_northing",0],AUTHORITY["EPSG","2927"],AXIS["X",EAST],AXIS["Y",NORTH]]')

// set up a fake db object with the methods the constructor is looking for
var db = {
  getWKT: function (wkid, callback) {
    const wkt = wkidHash.get(wkid)
    if (!wkt) return callback(null, undefined)
    callback(null, wkt)
  },
  insertWKT: function (wkid, wkt, callback) {
    wkidHash.set(wkid, wkt)
    callback(null)
  }
}

var sr = new SpatialReference({db: db})

test('e2e tests', function (t) {
  t.test('when the wkid is not in the db, not in ESRI, call epsg.io', function (t) {
    t.plan(4)
    sr.wkidToWkt(7088, function (err, wkt) {
      t.equals(err, null)
      t.equals(wkt, 'GEOGCS["GRS 1980(IUGG, 1980)",DATUM["unknown",SPHEROID["GRS80",6378137,298.257222101]],PRIMEM["Greenwich",0],UNIT["degree",0.0174532925199433],AUTHORITY["epsg","7088"]]')
      t.equals(err, null)
      t.equals(wkidHash.get(7088), 'GEOGCS["GRS 1980(IUGG, 1980)",DATUM["unknown",SPHEROID["GRS80",6378137,298.257222101]],PRIMEM["Greenwich",0],UNIT["degree",0.0174532925199433],AUTHORITY["epsg","7088"]]')
    })
  })

  t.test('when wkid not found anywhere - error', function (t) {
    t.plan(1)
    sr.wkidToWkt(9999, function (err, wkt) {
      t.ok(err, null)
    })
  })
})


test/index.js

const test = require('tape')
const SpatialReference = require('../src/index')
const wkidHash = new Map()
wkidHash.set(2927, 'PROJCS["NAD83(HARN) / Washington South (ftUS)",GEOGCS["NAD83(HARN)",DATUM["NAD83_High_Accuracy_Regional_Network",SPHEROID["GRS 1980",6378137,298.257222101,AUTHORITY["EPSG","7019"]],TOWGS84[-0.9956,1.9013,0.5215,0.025915,0.009426,0.011599,-0.00062],AUTHORITY["EPSG","6152"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4152"]],UNIT["US survey foot",0.3048006096012192,AUTHORITY["EPSG","9003"]],PROJECTION["Lambert_Conformal_Conic_2SP"],PARAMETER["standard_parallel_1",47.33333333333334],PARAMETER["standard_parallel_2",45.83333333333334],PARAMETER["latitude_of_origin",45.33333333333334],PARAMETER["central_meridian",-120.5],PARAMETER["false_easting",1640416.667],PARAMETER["false_northing",0],AUTHORITY["EPSG","2927"],AXIS["X",EAST],AXIS["Y",NORTH]]')

// set up a fake db object with the methods the constructor is looking for
var db = {
  getWKT: function (wkid, callback) {
    const wkt = wkidHash.get(wkid)
    if (!wkt) return callback(null, undefined)
    callback(null, wkt)
  },
  insertWKT: function (wkid, wkt, callback) {
    wkidHash.set(wkid, wkt)
    callback(null)
  }
}

var sr = new SpatialReference({db: db})

test('translating wkid to wkt', function (t) {
  t.test('when the wkid is in the db', function (t) {
    t.plan(2)
    sr.wkidToWkt(2927, function (err, wkt) {
      t.equals(err, null)
      t.equals(wkt, 'PROJCS["NAD83(HARN) / Washington South (ftUS)",GEOGCS["NAD83(HARN)",DATUM["NAD83_High_Accuracy_Regional_Network",SPHEROID["GRS 1980",6378137,298.257222101,AUTHORITY["EPSG","7019"]],TOWGS84[-0.9956,1.9013,0.5215,0.025915,0.009426,0.011599,-0.00062],AUTHORITY["EPSG","6152"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4152"]],UNIT["US survey foot",0.3048006096012192,AUTHORITY["EPSG","9003"]],PROJECTION["Lambert_Conformal_Conic_2SP"],PARAMETER["standard_parallel_1",47.33333333333334],PARAMETER["standard_parallel_2",45.83333333333334],PARAMETER["latitude_of_origin",45.33333333333334],PARAMETER["central_meridian",-120.5],PARAMETER["false_easting",1640416.667],PARAMETER["false_northing",0],AUTHORITY["EPSG","2927"],AXIS["X",EAST],AXIS["Y",NORTH]]')
    })
  })

  t.test('when the wkid is not in the db, but is in ESRI current', function (t) {
    t.plan(4)
    sr.wkidToWkt(4269, function (err, wkt) {
      t.equals(err, null)
      t.equals(wkt, 'GEOGCS["GCS_North_American_1983",DATUM["D_North_American_1983",SPHEROID["GRS_1980",6378137.0,298.257222101]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]]')
      t.equals(err, null)
      t.equals(wkidHash.get(4269), 'GEOGCS["GCS_North_American_1983",DATUM["D_North_American_1983",SPHEROID["GRS_1980",6378137.0,298.257222101]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]]')
    })
  })

  t.test('when the wkid is not in the db, but is in ESRI deprecated', function (t) {
    t.plan(4)
    sr.wkidToWkt(102645, function (err, wkt) {
      const actualWkt = 'PROJCS["NAD_1983_StatePlane_California_V_FIPS_0405_Feet",GEOGCS["GCS_North_American_1983",DATUM["D_North_American_1983",SPHEROID["GRS_1980",6378137.0,298.257222101]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]],PROJECTION["Lambert_Conformal_Conic"],PARAMETER["False_Easting",6561666.666666666],PARAMETER["False_Northing",1640416.666666667],PARAMETER["Central_Meridian",-118.0],PARAMETER["Standard_Parallel_1",34.03333333333333],PARAMETER["Standard_Parallel_2",35.46666666666667],PARAMETER["Latitude_Of_Origin",33.5],UNIT["Foot_US",0.3048006096012192]]'

      t.equals(err, null)
      t.equals(wkt, actualWkt)
      t.equals(err, null)
      t.equals(wkidHash.get(102645), actualWkt)
    })
  })
})

test('direct esri lookup', function (t) {
  t.test('when the wkid is not deprecated', function (t) {
    t.plan(2)
    sr.esriLookup(4269, function (err, wkt) {
      t.equals(err, null)
      t.equals(wkt, 'GEOGCS["GCS_North_American_1983",DATUM["D_North_American_1983",SPHEROID["GRS_1980",6378137.0,298.257222101]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]]')
    })
  })
})


src/index.js
const request = require('request')
const fixWkt = require('./fixwkt')
const esriCodes = require('@esri/proj-codes')

/**
 * Define the SpatialReference class
 */
class SpatialReference {
  constructor (options) {
    // If a db was passed in, ensure it has required methods, then add to instance
    if (options.db && options.db.getWKT && options.db.insertWKT) {
      this.db = options.db
    } else if (options.db) {
      throw new Error('Passed in db does not support getWKT or insertWKT')
    }
    // Add optional logger to instance
    this.logger = options.logger
  }

  /**
   * Convert a wkid to a wkt, by 1) looking in db, 2) check ESRI lookups, 3) asking epsg.io
   * @param {integer} wkid
   * @param {function} callback
   */
  wkidToWkt (wkid, callback) {
    // If a db used, check there first
    if (this.db) {
      getFromDB.call(this, wkid)
        .then(wkt => {
          if (wkt) {
            return callback(null, fixWkt(wkt))
          }
          // If wkt was not found in db, check Esri lookup
          wkt = getWktFromEsri(wkid)
          if (wkt) return wkt

          // If wkt was not found in Esri lookup, check EPSG
          return getFromApi(wkid)
        })
        .then(wkt => {
          // If the wkt
          if (!wkt) return

          // Insert into DB
          return this.db.insertWKT(wkid, wkt, (err) => {
            if (err) log.call(this, 'error', 'Failed to insert WKT into database: ' + wkid + ' ' + wkt + ', ' + err.message)
            // Log DB insertion error, but pass on wkt to callback
            callback(null, fixWkt(wkt))
          })
        })
        .catch(err => {
          log(this, 'error', err.message)
          return callback(err)
        })
    } else {
      // No db, look in Esri first
      const wkt = getWktFromEsri(wkid)

      // If wkt found in Esri, execute callback and end here
      if (wkt) return callback(null, fixWkt(wkt))

      // If not found in Esri, look in EPSG
      getFromApi(wkid)
        .then(wkt => {
          return callback(null, fixWkt(wkt))
        })
        .catch(err => {
          log(this, 'error', err.message)
          return callback(err)
        })
    }
  }

  /**
   * Wrapper around ESRI lookup
   * @param {integer} wkid
   * @param {function} callback
   */
  esriLookup (wkid, callback) {
    const wkt = getWktFromEsri(wkid)
    if (!wkt) {
      log.call(this, 'error', 'WKT not found for WKID: ' + wkid)
      return callback(new Error('WKT not found'))
    }
    callback(null, wkt)
  }
}

/**
 * Check the DB to see if it already contains the wkid
 * @param {integer} wkid
 */
function getFromDB (wkid) {
  return new Promise((resolve, reject) => {
    this.db.getWKT(wkid, function (err, wkt) {
      if (err) reject(err)
      if (!wkt) resolve()
      resolve(wkt)
    })
  })
}

/**
 * Request WKT for a wkid from epsg.io
 * @param {integer} wkid
 */
function getFromApi (wkid) {
  const url = 'http://epsg.io/' + wkid + '.wkt'
  return new Promise((resolve, reject) => {
    request.get(url, function (err, res, body) {
      if (err) reject(err)
      if (res.statusCode >= 400) reject(new Error(`Request to ${url} returned with status ${res.statusCode}.`))
      resolve(body)
    })
  })
}

/**
 * Look in current and deprecated ESRI projection codes
 * @param {integer} wkid
 */
function getWktFromEsri (wkid) {
  const result = esriCodes.lookup(wkid)
  if (result) return result.wkt
}

/**
 * Log messages
 * @param {string} level
 * @param {string} message
 */
function log (level, message) {
  if (!this.logger) return console.error(message)
  this.logger.log(level, message)
}

module.exports = SpatialReference


src/fixwkt.js
/**
 * Fixes incorrect WKT Strings
 *
 * @param {string} inWkt - The original well-known text for the projection
 * @param {integer} wkid - The well-known id for the projection
 * @private
 */
module.exports = function (inWkt, wkid) {
  // we have issue projecting this WKID w/o a datum xform
  // FYI there may be other proj codes needed here
  var wkt
  switch (wkid) {
    case 3078:
      wkt = '+proj=omerc +lat_0=45.30916666666666 +lonc=-86 +alpha=337.25556 +k=0.9996 +x_0=2546731.496 +y_0=-4354009.816 + ellps=GRS80 +datum=NAD83 +units=m +no_defs'
      break
    case 28992:
      wkt = '+title=Amersfoort/Amersfoort +proj=sterea +lat_0=52.15616055555555 +lon_0=5.38763888888889 +k=0.999908 +x_0=155000 +y_0=463000 +ellps=bessel +units=m +no_defs +towgs84=565.2369,50.0087,465.658,-0.406857330322398,0.350732676542563,-1.8703473836068,4.0812'
      break
    case 5514:
      wkt = 'PROJCS["S-JTSK_Krovak_East_North",GEOGCS["GCS_S_JTSK",DATUM["Jednotne_Trigonometricke_Site_Katastralni",SPHEROID["Bessel_1841",6377397.155,299.1528128]],TOWGS84[589,76,480,0,0,0,0],PRIMEM["Greenwich",0],UNIT["Degree",0.017453292519943295]],PROJECTION["Krovak"],PARAMETER["False_Easting",0],PARAMETER["False_Northing",0],PARAMETER["Pseudo_Standard_Parallel_1",78.5],PARAMETER["Scale_Factor",0.9999],PARAMETER["Azimuth",30.28813975277778],PARAMETER["Longitude_Of_Center",24.83333333333333],PARAMETER["Latitude_Of_Center",49.5],PARAMETER["X_Scale",-1],PARAMETER["Y_Scale",1],PARAMETER["XY_Plane_Rotation",90],UNIT["Meter",1],AUTHORITY["EPSG","102067"]]'
      break
    case 2264:
      wkt = 'PROJCS["NAD_1983_StatePlane_North_Carolina_FIPS_3200_Feet",GEOGCS["GCS_North_American_1983",DATUM["D_North_American_1983",SPHEROID["GRS_1980",6378137.0,298.257222101]],TOWGS84[-0.99343,1.90331,0.52655,0.02591467,0.00942645,0.01159935,-0.00171504],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]],PROJECTION["Lambert_Conformal_Conic"],PARAMETER["False_Easting",2000000.002616666],PARAMETER["False_Northing",0.0],PARAMETER["Central_Meridian",-79.0],PARAMETER["Standard_Parallel_1",34.33333333333334],PARAMETER["Standard_Parallel_2",36.16666666666666],PARAMETER["Latitude_Of_Origin",33.75],UNIT["Foot_US",0.3048006096012192]]'
      break
    default:
      wkt = inWkt
  }
  // per Melita Kennedy @esri This is the correct datum transformation to NAD83 for the variant of WGS84 used in recent data. See also: http://www.epsg-registry.org/

  if (wkt.match(/NAD83\(HARN\)/)) return wkt.replace('TOWGS84[0,0,0,0,0,0,0]', 'TOWGS84[-0.9956,1.9013,0.5215,0.025915,0.009426,0.011599,-0.00062]')
  else if (wkt.match(/NAD83/)) return wkt.replace('TOWGS84[0,0,0,0,0,0,0]', 'TOWGS84[-0.99343,1.90331,0.52655,0.02591467,0.00942645,0.01159935,-0.00171504]')
  else if (wkt.match(/UTM/)) return wkt.replace('TOWGS84[0,0,0,0,0,0,0]', 'TOWGS84[-0.9956,1.9013,0.5215,0.025915,0.009426,0.011599,-0.00062]')
  return wkt
}
